src/btree/bt_cursor.c

```diff
@@ -254,17 +254,22 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint64_t recno, bool *vali
     switch (btree->type) {
     case BTREE_COL_FIX:
         /*
-         * If search returned an insert object, there may or may not be a matching on-page object,
-         * we have to check. Fixed-length column-store pages don't have slots, but map one-to-one to
-         * keys, check for retrieval past the end of the page.
+         * If search returned an insert, we might be past the end of page in the append list, so
+         * there's no on-disk value.
          */
         if (cbt->recno >= cbt->ref->ref_recno + page->entries)
             return (0);
 
-        *valid = true;
         /*
-         * An update would have appeared as an "insert" object; no further checks to do.
+         * Check for an update ondisk or in the history store. For column store, an insert object
+         * can have the same key as an on-page or history store object.
          */
+        WT_RET(__wt_txn_read(session, cbt, key, recno, cbt->ins ? cbt->ins->upd : NULL));
+        if (cbt->upd_value->type != WT_UPDATE_INVALID) {
+            if (cbt->upd_value->type == WT_UPDATE_TOMBSTONE)
+                return (0);
```

Comment:
Because there's code for that downstream (look for __cursor_fix_implicit) -- I had it the other way here originally, and it doesn't work, although I don't remember how. (I was very suspicious and left the change uncommitted for like three weeks so I could revert it on demand if something broke, but nothing ever did; but that means the consequences were a while ago.)

Reply:
I think this one needs a comment, please.

Reply:
Fixed.